yearn/multicall2.py

```diff
@@ -21,18 +21,25 @@
 def fetch_multicall(*calls, block=None, require_success=False):
     # https://github.com/makerdao/multicall
     multicall_input = []
+    attribute_errors = []
     fn_list = []
     decoded = []
 
-    for contract, fn_name, *fn_inputs in calls:
-        fn = getattr(contract, fn_name)
+    for i, (contract, fn_name, *fn_inputs) in enumerate(calls):
+        try:
+            fn = getattr(contract, fn_name)
 
-        # check that there aren't multiple functions with the same name
-        if hasattr(fn, "_get_fn_from_args"):
-            fn = fn._get_fn_from_args(fn_inputs)
+            # check that there aren't multiple functions with the same name
+            if hasattr(fn, "_get_fn_from_args"):
+                fn = fn._get_fn_from_args(fn_inputs)
 
-        fn_list.append(fn)
-        multicall_input.append((contract, fn.encode_input(*fn_inputs)))
+            fn_list.append(fn)
+            multicall_input.append((contract, fn.encode_input(*fn_inputs)))
+        except AttributeError:
+            if not require_success:
+                attribute_errors.append(i)
+                continue
+            raise
```

Comment:
not sure I understand what this is doing / which error you've observed, maybe we could have a chat

Reply:
lets say for example we are fetching price for a yearn vault

in pseudocode, you do something like this

fetch_multicall([decimals,ppfs,underlying[,require_success=False)

with require_success=False, if a contract doesn't have a specific method we would expect fetch_multicall to return None for that method. But before the fix, you would get an AttributeError. This change solves for that case.

Reply:
ok, but then don't do L68-L69 but instead handle it already in the rescue block:
`decoded.append(None)` like in L65 and get rid of `attribute_errors` ?

Reply:
If we do that we would get the outputs out of order, I used the cached index i to slot the Nones into place after the multicall